src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/pool-edit-mode-modal/pool-edit-mode-modal.component.ts

```diff
@@ -0,0 +1,83 @@
+import { Component, OnInit } from '@angular/core';
+import { FormControl, Validators } from '@angular/forms';
+
+import { BsModalRef } from 'ngx-bootstrap/modal';
+
+import { RbdMirroringService } from '../../../../shared/api/rbd-mirroring.service';
+import { CdFormGroup } from '../../../../shared/forms/cd-form-group';
+import { ExecutingTask } from '../../../../shared/models/executing-task';
+import { FinishedTask } from '../../../../shared/models/finished-task';
+import { TaskWrapperService } from '../../../../shared/services/task-wrapper.service';
+import { PoolEditModeResponseModel } from './pool-edit-mode-response.model';
+
+@Component({
+  selector: 'cd-pool-edit-mode-modal',
+  templateUrl: './pool-edit-mode-modal.component.html',
+  styleUrls: ['./pool-edit-mode-modal.component.scss']
+})
+export class PoolEditModeModalComponent implements OnInit {
+  poolName: string;
+  executingTasks: ExecutingTask[];
+
+  editModeForm: CdFormGroup;
+  bsConfig = {
+    containerClass: 'theme-default'
+  };
+  pattern: string;
+
+  response: PoolEditModeResponseModel;
+
+  mirrorModes: Array<{ id: string; name: string }> = [
+    { id: 'disabled', name: 'Disabled' },
+    { id: 'pool', name: 'Pool' },
+    { id: 'image', name: 'Image' }
+  ];
+
+  constructor(
+    public modalRef: BsModalRef,
+    private rbdMirroringService: RbdMirroringService,
+    private taskWrapper: TaskWrapperService
+  ) {
+    this.createForm();
+  }
+
+  createForm() {
+    this.editModeForm = new CdFormGroup({
+      mirrorMode: new FormControl('', {
+        validators: [Validators.required]
+      })
+    });
+  }
+
+  ngOnInit() {
+    this.pattern = `${this.poolName}`;
+    this.rbdMirroringService.getPool(this.poolName).subscribe((resp: PoolEditModeResponseModel) => {
+      this.setResponse(resp);
+    });
+  }
+
+  setResponse(response: PoolEditModeResponseModel) {
+    this.editModeForm.get('mirrorMode').setValue(response.mirror_mode);
+  }
+
+  update() {
+    const request = new PoolEditModeResponseModel();
+    request.mirror_mode = this.editModeForm.getValue('mirrorMode');
+
+    const action = this.taskWrapper.wrapTaskAroundCall({
+      task: new FinishedTask('rbd/mirroring/pool/edit', {
+        pool_name: this.poolName
+      }),
+      call: this.rbdMirroringService.updatePool(this.poolName, request)
+    });
+
+    action.subscribe(
+      undefined,
+      () => this.editModeForm.setErrors({ cdSubmitButton: true }),
+      () => {
+        this.rbdMirroringService.refresh();
```

Comment:
Is there a proper way to invoke a callback once the API call has completed successfully? I want to ensure that the UI gets properly updated w/o the need to click the refresh buttons. Right now this seems racy in that sometimes it works and sometimes it doesn't.

Reply:
`refresh()` needs to return an Observable.

Changing `refresh `to the following should work:
```
refresh() {
  this.http
    .get('api/block/mirroring/summary')
    .pipe(tap((data) => this.summaryDataSource.next(data)));
}
```

then you need to subscribe after each call:
```suggestion
this.rbdMirroringService.refresh().subscribe((data) => {
  //It's finished
});
```

Reply:
... it turns out that the controller endpoint for "/api/block/mirroring/summary" will cache the view for up to 1 second, which is why I am sometimes getting stale results. I guess that cache needs to be removed or I need a way to force-refresh.